thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java

```diff
@@ -414,6 +416,70 @@ public Response updateDetectionPipelineApi(
     return Response.ok().entity(responseMessage).build();
   }
 
+  private DetectionConfigDTO fetchExistingDetection(String payload) {
+    DetectionConfigDTO existingDetectionConfig = null;
+
+    // Extract the detectionName from payload
+    Map<String, Object> detectionConfigMap = new HashMap<>();
+    detectionConfigMap.putAll(ConfigUtils.getMap(this.yaml.load(payload)));
+    String detectionName = MapUtils.getString(detectionConfigMap, PROP_DETECTION_NAME);
+    Preconditions.checkNotNull(detectionName, "Missing property detectionName in the detection config.");
+
+    // Check if detection already existing
+    Collection<DetectionConfigDTO> detectionConfigs = this.detectionConfigDAO
+        .findByPredicate(Predicate.EQ("name", detectionName));
+    if (detectionConfigs != null && !detectionConfigs.isEmpty()) {
+      existingDetectionConfig = detectionConfigs.iterator().next();
+    }
+
+    return existingDetectionConfig;
+  }
+
+  long createUpdateDetectionPipeline(String payload) {
+    Preconditions.checkArgument(StringUtils.isNotBlank(payload), "The Yaml Payload in the request is empty.");
+    long detectionId;
+    DetectionConfigDTO existingDetection = fetchExistingDetection(payload);
+    if (existingDetection != null) {
+      detectionId = existingDetection.getId();
+      updateDetectionPipeline(detectionId, payload, 0, 0);
+    } else {
+      detectionId = createDetectionPipeline(payload, 0, 0);
+    }
+
+    return detectionId;
+  }
+
+  /**
+   Set up a detection pipeline using a YAML config - create new or update existing
+   @param payload YAML config string
+   @return a message contains the saved detection config id
+   */
+  @POST
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.TEXT_PLAIN)
+  @ApiOperation("Create a new detection pipeline or update existing if one already exists")
+  public Response createUpdateDetectionPipelineApi(@ApiParam("yaml config") String payload) {
+    Map<String, String> responseMessage = new HashMap<>();
+    long detectionConfigId;
+    try {
+      detectionConfigId = createUpdateDetectionPipeline(payload);
+    } catch (IllegalArgumentException e) {
```

Comment:
The catch blocks are duplicated in YamlResource. Suggest to put them together.

Reply:
@xiaohui-sun, I didn't quite get the point. Can you provide some details?

Reply:
There are multiple places to handle pipeline creation exceptions. If you search for "Validation error while creating subscription group with payload" there are already 2 matches. This is a minor  .

Reply:
I have addressed it in a separate PR: https://github.com/apache/incubator-pinot/commit/1ff23c3628e2148910da940d1a84afb791a9a3d9